packages/peregrine/lib/talons/FilterModal/useFilterModal.js

```diff
@@ -0,0 +1,176 @@
+
+            // map item values to items
+            for (const value of params.getAll(key)) {
+                items.add(groupItemsByValue.get(value));
+            }
+
+            // add items to the next state, keyed by group
+            nextState.set(group, items);
+        }
+    }
+
+    return nextState;
+};
+
+const getSearchFromState = (initialValue, filterKeys, filterState) => {
+    // preserve all existing params
+    const nextParams = new URLSearchParams(initialValue);
+
+    // iterate over available filters
+    for (const key of filterKeys) {
+        // remove any existing filter values
+        nextParams.delete(key);
+    }
+
+    // iterate over the latest filter values
+    for (const [group, items] of filterState) {
+        for (const item of items) {
+            const { title, value } = item || {};
+
+            // append the new values
+            nextParams.append(`${group}[title]`, `${title}`);
+            nextParams.append(`${group}[value]`, `${value}`);
+        }
+    }
+
+    // prepend `?` to the final string
+    return `?${nextParams.toString()}`;
+};
+
+/**
+ * Filter Modal talon.
+ *
+ * @returns {{
+ *   filterApi: any,
+ *   filterState: any,
+ *   handleClose: function,
+ *   isOpen: boolean
+ * }}
+ */
+export const useFilterModal = filters => {
+    const [isApplying, setIsApplying] = useState(false);
+    const [{ drawer }, { closeDrawer }] = useAppContext();
+    const [filterState, filterApi] = useFilterState();
+    const prevDrawer = useRef(null);
+    const isOpen = drawer === 'filter';
+
+    const history = useHistory();
+    const { pathname, search } = useLocation();
+
+    // iterate over filters once to set up all the collections we need
+    const [filterNames, filterKeys, filterItems] = useMemo(() => {
+        const names = new Map();
+        const keys = new Set();
+        const itemsByGroup = new Map();
+
+        for (const filter of filters) {
+            const { filter_items, name, request_var: group } = filter;
+            const items = [];
+
+            // add filter name
+            names.set(group, name);
+
+            // add filter key permutations
+            keys.add(`${group}[title]`);
+            keys.add(`${group}[value]`);
+
+            // add items
+            for (const { label, value_string: value } of filter_items) {
+                items.push({ title: stripHtml(label), value });
+            }
+            itemsByGroup.set(group, items);
+        }
+
+        return [names, keys, itemsByGroup];
+    }, [filters]);
+
+    // on apply, write filter state to history and close the drawer
+    useEffect(() => {
+        if (isApplying) {
+            const nextSearch = getSearchFromState(
+                search,
+                filterKeys,
+                filterState
+            );
+
+            // write filter state to history
+            history.push({ pathname, search: nextSearch });
+            // close the drawer
+            closeDrawer();
```

Comment:
Why is this effect a thing and not just what the `handleApply` callback does?

Reply:
Well, a callback that did all of those things would end up being recreated whenever `filterState` changed. This way, the effect has the latest dependencies, but the callback never changes.

I do wonder about this pattern though. Do you think it's too indirect?

Reply:
Aside from readability I'm not sure there's a drawback to this method. It's just tough to parse at first but I guess the alternative would be to pass state directly to the callback which means _something_ has to be aware of state.

Reply:
This pattern really is necessary. There are alternatives, such as passing state around or accessing it through a ref, but I think this is the cleanest one so far. We can always revisit, since this isn't a totally unique case.